walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -117,4 +129,67 @@ public int compare(Event a, Event b) {
 
   return result; 
   }
+
+  private Collection<TimeRange> optionalQuery(Collection<Event> events, MeetingRequest request) {
+    
+    // Find out how long they have to be at the meeting
+    long duration = request.getDuration(); 
+    // Set the collection of possible times to be empty
+    Collection<TimeRange> result = new ArrayList<TimeRange>(); 
+
+    // window referring to the opening and closing of the window of time where the meeting could be done
+    int windowStart = 0; 
+    int windowClose = 0; 
+
+    Object[] objectArray = events.toArray(); 
+    Event[] eventsOrderedByStart = new Event[objectArray.length]; 
+    Event[] eventsOrderedByEnd = new Event[objectArray.length];
+
+    for (int i = 0; i < events.toArray().length; i++) {
+      eventsOrderedByStart[i] = (Event) objectArray[i];  
+      eventsOrderedByEnd[i] = (Event) objectArray[i]; 
+    }
+
+    Arrays.sort(eventsOrderedByStart, ORDER_EVENT_BY_START);
+    Arrays.sort(eventsOrderedByEnd, ORDER_EVENT_BY_END);
+    int startPointer = 0; 
+    int endPointer = 0; 
+    int nextMeetingEnd = 0; 
+    int nextMeetingStart = 0; 
+    Set<Event> problemEvents = new HashSet<>();
+
+
+
+    // Loop until all meetings have ended
+    while(endPointer < eventsOrderedByEnd.length) {
+    // What comes first, the end of a meeting or the start of one
+      nextMeetingEnd = eventsOrderedByEnd[endPointer].getWhen().end(); 
+      if (startPointer < eventsOrderedByStart.length) {
+      nextMeetingStart = eventsOrderedByStart[startPointer].getWhen().start(); 
+      }
+      if (startPointer == eventsOrderedByStart.length || nextMeetingEnd <= nextMeetingStart) {
+        if (problemEvents.contains(eventsOrderedByEnd[endPointer])) {
+          windowStart = nextMeetingEnd; 
+          problemEvents.remove(eventsOrderedByEnd[endPointer]); 
+        }
+        endPointer++; 
```

Comment:
```
.
.
endPointer++; 
continue;
}
if (isEventImportantWithOptional(eventsOrderedByStart[startPointer], request)) {
            windowClose = nextMeetingStart; 
            if (problemEvents.isEmpty()) {
              if ((windowClose - windowStart) >= duration) {
                result.add(TimeRange.fromStartEnd(windowStart, windowClose, false)); 
              }
            }
            problemEvents.add(eventsOrderedByStart[startPointer]); 
          }
          startPointer++; 
```

Reply:
reduce the if-else complexity whenever possible

Reply:
Okay, I just finished a newer version of this section of the code, and it's still overly complicated, so I'll work on that this afternoon.